### ISSUE TITLE
Wrong result when adding an inconsequential variable

### ISSUE THREAD

**Issue by CelineWuest** [neo4j/neo4j#13487]
I found a discrepancy when running two semantically equivalent queries against an empty database:

---
1:
```cypher
CREATE ()-[:A]->()
------------------------
OPTIONAL MATCH ()
OPTIONAL MATCH () WHERE false
RETURN COUNT(*) AS n1
	UNION ALL
CREATE ()
RETURN 0 AS n1
```
Produces:
```
╒═══╕
│n1 │
╞═══╡
│0  │
├───┤
│3  │
└───┘
```

---
2:
```cypher
CREATE ()-[:A]->()
------------------------
OPTIONAL MATCH ()
WITH 0 AS n0
OPTIONAL MATCH () WHERE false
RETURN COUNT(*) AS n1
	UNION ALL
CREATE ()
RETURN 0 AS n1
```
Produces:
```
╒═══╕
│n1 │
╞═══╡
│0  │
├───┤
│2  │
└───┘
```
---

I believe the queries mentioned above are semantically equivalent and thus their result should be the same.

I encountered this issue when testing queries against the **Neo4j 5.21 enterprise version** in a Docker container running **alpine v.3**.

### Steps to reproduce
Run the aforementioned, semantically equivalent, queries against an empty database and observe that their results differ.

### Expected behavior
The queries should result in the same behavior

### Actual behavior
The query results differ

**Comment by SachinVasant** [comment]
Hi @DominicWuest ,
Thank you so much for reporting this issue. We will investigate it further and keep you posted.

**Comment by henriknyman** [comment]
@DominicWuest Thanks for reporting! I have merged a fix that should be included in the 5.24 release.
